### DISCUSSION TITLE
Mark certain commands to allow expansion of regular abbreviations following them

### DISCUSSION THREAD

**Discussion by henrebotha** [olets/zsh-abbr#99] [Ideas] [answered]
Certain shell commands, such as `sudo` or `watch`, are typically immediately followed by a full shell command. One might conceptualise them not as commands in their own right, but more as modifiers to commands. I don't for example think of `sudo apt install` as a separate command to `apt install`, but rather as applying `sudo` as a modifier to `apt install`.

I would like to be able to indicate to zsh-abbr that a given command is to be understood as preceding a full command, and thus that any **regular** (non-global) abbreviations must be expanded after that command.

So if I have `k gp` as an abbreviation of `kubectl get pods`, and I mark `sudo` and `watch` as "modifier commands", then:

- `watch k gp` should expand to `watch kubectl get pods`;
- `sudo k gp` should expand to `sudo kubectl get pods`;
- `echo k gp` should not expand to anything.

This behaviour is possible with great manual effort, but it would be a lot more convenient if zsh-abbr supported it directly as a feature.

**Comment by olets** [comment]
Interesting idea, thanks for the suggestion.

The way to support this that comes to mind is an array of strings to ignore at the front of the buffer. Default empty array. Call it "command expansion ignored prefixes" or something like that. Loop over the array, trimming at most one of those strings off the front of the buffer and then continuing out of the loop. Haven't yet spent the time to think about the "can be done but should it?" question.

//


> great manual effort

Enhancing zsh-abbr is a low priority for me currently, so I wonder if we can come up with a solution for your need that doesn't require zsh-abbr enhancement.

I don't know your workflows of course, but on the surface it looks like you could

```sh
abbr -g "k gp"="kubectl get pods"
```

and get the echo example to work with

```sh
echo k gp;
```

or

```sh
abbr -g k=kubectl
abbr -g "kubectl gp"="kubectl get pods"
```

and then

```sh
echo k<ctrl space>gp
```

**Comment by henrebotha** [comment]
Yeah I have a solution that is sufficient for now. I have a section in my dotfiles that automatically sets up certain abbreviations on startup (it's kind of silly since I can just stick my abbreviations file directly in my dotfiles, but I'll do that some other day lol). I have added [this part](https://github.com/henrebotha/dotfiles/commit/e8f11aa5ef93e8f7824bea2fcc8deeda4bd3b59a) to it to iterate over all my desired abbreviations and create variations of them prefixed with `sudo` and `watch` respectively. I haven't used it for long enough to figure out what issues may exist with this approach, but at least it satisfies the naive version of my request.

**Comment by olets** [comment, marked as answer]
- Superseded by https://github.com/olets/zsh-abbr/discussions/114
